### PR TITLE
ReflectometryReductionOneAuto code tidied

### LINKED ISSUE #26499: ReflectometryReductionOneAuto tidying
As noted by @DanNixon on #25436 it would be useful to do some tidying in `ReflectometryReductionOneAuto3`. Initial suggestions are:

- [x] Simplify the `RebinParams` class
- [x] Remove extraneous newlines after function comments
- [x] Replace `catch (...)` with something more explicit and/or check if the workspace exists first.
- [x] Make use of east/west const consistent
- [x] In `validateInputs`, there is duplicated code for the checks for first/second transmission runs - refactor this into a function

We should also review the rest of this file as part of #23012 to see if more tidying can be done.

### PR DESCRIPTION
**Pull request by giovannidisiena**
**Description of work.**
- `RebinParams` class simplified
- Replaced `catch (...)` with check to see if workspace exists
- Made west const consistent
- Refactored duplicated transmission run code in `validateInputs` into a function
- Added check to prevent unhandled exception when comparing informative/input workspace names

**To test:**
Code review.

We should also do some manual testing as this PR makes changes to functional code relating to re-binning, transmission run validation and slicing.

1. - Open `ISIS Reflectometry` interface
2. - Change instrument to `OFFSPEC`
3. - Enter run number `44956`, angle `0.4` and first trans run `44937`
4. - Process and plot
5. - Alter `QMin`/`QMax`/`dQ/Q` values
6. - Process and plot again

- Delete all workspaces

1. - Change instrument to `INTER`
2. - Enter run number `13460` and angle `0.5`
3. - Click on `Event Handling` tab
4. - Enable `Uniform slicing`
5. - Increase number of slices to `2`
6. - Process and observe output workspaces are correct

Closes #26499 

*This does not require release notes as there is no change in functionality*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by gemmaguest — CHANGES_REQUESTED** [review]
Code changes look good :+1: Just a minor suggestion for the new test. We should also do some manual testing as this PR makes changes to functional code. Could you have a think about what would be good to test based on the code that is touched? My first thought is to test with/without polarisation corrections - a good run for that is `44956` with angle `0.4` and transmission run `44937`.

**Comment by giovannidisiena** [comment]
@gemmaguest I have update the workspace naming however I am not too sure about the functional testing, may be worth discussing later 🙂

**Comment by giovannidisiena** [comment]
Tests pass for me locally on Windows however I get this on the Ubuntu VM... any idea as to why @gemmaguest?

RROA3_test_fail

**Comment by gemmaguest** [comment]
> Tests pass for me locally on Windows however I get this on the Ubuntu VM... any idea as to why @gemmaguest?
> 
> RROA3_test_fail

It may be that the processing instructions are for a spectrum that isn't a valid detector. Although if that's the case I'm surprised it works on Windows.

File: Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h [hunk 1]
```diff
@@ -44,29 +44,22 @@ class DLLExport ReflectometryReductionOneAuto3
     std::string iVsLam;
   };
 
-  class RebinParams {
-  public:
+  struct RebinParams {
     RebinParams(const double qMin, const bool qMinIsDefault, const double qMax,
```
Comment (gemmaguest) [inline Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]: You can simplify this further by removing the constructor - it isn't required if you use an initializer list when you create an instance (e.g. see [here](https://docs.microsoft.com/en-us/cpp/cpp/initializing-classes-and-structs-without-constructors-cpp?view=vs-2019)).

File: Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h [hunk 2]
```diff
@@ -44,29 +44,22 @@ class DLLExport ReflectometryReductionOneAuto3
...
-    double qMax() const { return m_qMax; };
-    bool qMaxIsDefault() const { return m_qMaxIsDefault; }
-    double qStep() const { return *m_qStep; };
-    bool hasQStep() const { return m_qStep.is_initialized(); }
-    std::vector<double> asVector() const {
-      return std::vector<double>{qMin(), qStep(), qMax()};
-    };
-
-  private:
     double m_qMin;
```
Comment (gemmaguest) [inline Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]: I think it's a more common convention to drop the `m_` prefix for public members - it makes it more readable when using them.

**Review by gemmaguest — CHANGES_REQUESTED** [review]
Just a couple more suggestions for simplifications

**Review by gemmaguest — DISMISSED** [review]
Code changes look good and the interface still functions as expected - rebinning to specified Q values works as does slicing and polarization correction :+1:

File: Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h [hunk 3]
```diff
@@ -91,31 +79,34 @@ class DLLExport ReflectometryReductionOneAuto3
...
                               const double defaultValue, bool &isDefault);
-  void setOutputWorkspaces(WorkspaceNames const &outputGroupNames,
+  void setOutputWorkspaces(const WorkspaceNames &outputGroupNames,
                            std::vector<std::string> &IvsLamGroup,
                            std::vector<std::string> &IvsQBinnedGroup,
                            std::vector<std::string> &IvsQGroup);
   WorkspaceNames getOutputNamesForGroups(const std::string &inputName,
                                          const std::string &runNumber,
                                          const size_t wsGroupNumber);
+  void getTransRun(std::map<std::string, std::string> &results,
```
Comment (DanNixon) [inline Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]:
```suggestion
  void getTransmissionRun(std::map<std::string, std::string> &results,
```

There is no real need to abbreviate this here.

File: Framework/Algorithms/src/ReflectometryReductionOneAuto3.cpp [hunk 4]
```diff
@@ -109,6 +109,34 @@ const std::string ReflectometryReductionOneAuto3::summary() const {
...
+/** Validate individual transmission runs
+ *
+ * @return :: void
+ */
+void ReflectometryReductionOneAuto3::getTransRun(
+    std::map<std::string, std::string> &results,
+    WorkspaceGroup_sptr &workspaceGroup, const std::string &transRun) {
+  const std::string str = getPropertyValue(transRun);
+  if (!str.empty()) {
+    auto transGroup =
```
Comment (DanNixon) [inline Framework/Algorithms/src/ReflectometryReductionOneAuto3.cpp]:
```suggestion
    auto transmissionGroup =
```

**Review by DanNixon — CHANGES_REQUESTED** [review]
Looks good, just a couple of minor comments.

File: Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h [hunk 5]
```diff
@@ -44,29 +44,17 @@ class DLLExport ReflectometryReductionOneAuto3
...
+    bool qMaxIsDefault;
+    boost::optional<double> qStep;
 
-    double qMin() const { return m_qMin; };
-    bool qMinIsDefault() const { return m_qMinIsDefault; }
-    double qMax() const { return m_qMax; };
-    bool qMaxIsDefault() const { return m_qMaxIsDefault; }
-    double qStep() const { return *m_qStep; };
-    bool hasQStep() const { return m_qStep.is_initialized(); }
+    bool hasQStep() const { return qStep.is_initialized(); }
```
[the remark this answers is not in the document]
Comment (giovannidisiena) [reply Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]: Could you explain this for me please?

[the remark this answers is not in the document]
Comment (DanNixon) [reply Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]: `boost::optional` (and `std::optional` for when we can use that) have `operator bool`. This allows conversion to `bool` when appropriate.

In the case of `optional` this returns if the optional has a value. The fact it is being returned from a function with a return type of `bool` indicates that `operator bool` should be used here.

[the remark this answers is not in the document]
Comment (giovannidisiena) [reply Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]: Thank you!

[the remark this answers is not in the document]
Comment (giovannidisiena) [reply Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]: @DanNixon now getting this error, any ideas?

`error: cannot convert ‘const boost::optional<double>’ to ‘bool’ in return`

[the remark this answers is not in the document]
Comment (DanNixon) [reply Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto3.h]: No idea. That should work. Just keep it as it was for now.

**Review by DanNixon — APPROVED** [review]

**Comment by DanNixon** [comment]
@giovannidisiena Maybe you could try and teach people how to tidy commits in the manner you did for the last three in this PR.

I've been fighting a losing battle trying to get people to actually do this.

**Pull request merged**